### PR TITLE
add generic bus interface for setting log level that can be implemented by any daemon

### PR DESCRIPTION
**Pull request by poettering**
result of the discussion from #15495

I quite like it

### CODE REVIEW

**Comment by boucman** [comment]
nice... could we have a sd_notify() way of doing that too ? or would that be a totally different codepath ?

**Comment by boucman** [comment]
Ok, disregard my comment, I think I misunderstood your patch... this is not about the daemon setting its DefaultLogLevel, it's about an external client asking the daemon to change its log level...

note that systemd already defines  an interface called org.freedesktop.systemd1.Service 

That's not exactly the same name, so it's fine... but  it might bring some confusion

(also: the new interface needs a man page... not sure where since the D-Bus man pages are per-client so far and that interface is not linked to a client)

**Comment by poettering** [comment]
> Ok, disregard my comment, I think I misunderstood your patch... this is not about the daemon setting its DefaultLogLevel, it's about an external client asking the daemon to change its log level...

Hmm, what were you looking for, what do you mean by DefaultLoglevel=? Note thta there is the SYSTEMD_LOG_LEVEL env var, at least for our own daemons, which can do something similar to this dbus interface, except ahead-of-time when starting a daemon.

> note that systemd already defines an interface called org.freedesktop.systemd1.Service

hmm, good point. I guess service is overloaded by many subsystems, i.e. dbus too. But I still would probably stick to the org.freedesktop.service1 name I must say...

> (also: the new interface needs a man page... not sure where since the D-Bus man pages are per-client so far and that interface is not linked to a client)

Yeah, definitely. but maybe let's agree on that this is a good thing first, before we add a man page...

**Comment by boucman** [comment]
I was thinkig of SysLogLevel= on services... so totally different, don't worry about it.

As for the interface name, there is a small question of ownership... technically all interfaces starting with org.freedesktop "belong" to freedesktop.org and the systemd project can't guarantee that there is no collision (interface name should be unique world-wide, that's why they start with a reverse DNS)

So systemd should stay in its sub-namespace org.freedesktop.systemd1, but we already have an interface named .Service there...

We could use io.systemd.service1 maybe ?

**Comment by poettering** [comment]
> So systemd should stay in its sub-namespace org.freedesktop.systemd1, but we already have an interface named .Service there...
> 
> We could use io.systemd.service1 maybe ?

My goal is explicitly to keep the door open for this to be something generic *any* daemon can implement, regardless if shipped with systemd or anywhere else. For example, would love if gdbus based daemons would start supporting this, too. Hence I picked a generic, non-systemd-specific name for this.

**Comment by boucman** [comment]
I disagree on that one... The fact that there is systemd in the name of an interface means that the interface is defined by systemd, not that only systemd can implement it... 

if I understand correctly the D-Bus specification, that's the very idea with the reverse-DNS naming convention of interfaces : interfaces are named after the one that defines/document them not the ones using them (even if they usually are the same...)

the only thing you seem to enforce in your patch (not 100% sure, I just quickly glanced at the code) is the object name: "/org/freedesktop/service1"
my understanding of dbus is that clients could use that interface anywhere it makes sense in their existing object hierarchy, not just in that object. So it might not be a good idea to enforce that. That should be a decision each daemon takes...

**Comment by poettering** [comment]
> I disagree on that one... The fact that there is systemd in the name of an interface means that the interface is defined by systemd, not that only systemd can implement it...

It's a political choice: if systemd is not in the name it makes it easier for others to defend implementing it unconditionally.

> if I understand correctly the D-Bus specification, that's the very idea with the reverse-DNS naming convention of interfaces : interfaces are named after the one that defines/document them not the ones using them (even if they usually are the same...)

Yupp, but we stepped into the fdo territory there, by using that prefix. I mean, the fdo prefix is kinda open range for anyone who wants stuff to be "generic" at this point. There's no body that controls what can be labelled fdo and what cannot be labelled that.

> the only thing you seem to enforce in your patch (not 100% sure, I just quickly glanced at the code) is the object name: "/org/freedesktop/service1"

Hmm? I declare an interface name, some properties in it, and an object path where to find it. That's all.

> my understanding of dbus is that clients could use that interface anywhere it makes sense in their existing object hierarchy, not just in that object. So it might not be a good idea to enforce that. That should be a decision each daemon takes...

Well, if clients don#t know on which object to find the interface there's no point in standardizing this.

Note that log levels are usually implemented as a per-service, global singleton, hence it makes sense to expose it as such, i.e. under a single, fixed object paths where you can reach it on any service that implements this "spec".

File: src/home/homed-manager.c [hunk 1]
```diff
@@ -894,6 +895,10 @@ static int manager_connect_bus(Manager *m) {
         if (r < 0)
                 return log_error_errno(r, "Failed to add object manager: %m");
 
+        r = bus_service_api_register(m->bus);
+        if (r < 0)
+                return r;
```
Comment (keszybz) [inline src/home/homed-manager.c:900]: Yeah, this is nice.

**Comment by boucman** [comment]
Ok, for the political aspect. I can get that...

I don't like the "free range" argument on fdo, especially if we take a term as generic as "service1" maybe instead of tying it to the idea of service, tie it to the idea of log management ? just like Introspectable is tied to the idea of... well introspectable 

As for the object, most daemons already have a singleton object that represent "them" and would be a more logical place to to put the interface (from the daemon's point of view, at least)

There is no standard dbus object for "the singleton representing this client", I know... but if such a thing is to be defined, again tying it to the term "service" might not be the best possible name.

**Comment by keszybz** [comment]
With documentation, I would prefer to wait until the vtable exporting things are done. Doesn't make much sense to do this now, and rework later.

**Comment by keszybz** [comment]
> tie it to the idea of log management

See the inline comments. More stuff like reloading and version reporting could be added later, going outside of the logging area.

**Comment by poettering** [comment]
> Ok, for the political aspect. I can get that...
> 
> I don't like the "free range" argument on fdo, especially if we take a term as generic as "service1" maybe instead of tying it to the idea of service, tie it to the idea of log management ? just like Introspectable is tied to the idea of... well introspectable

Hmm, so my thinking was that this would be something we would later extend for other service stuff too. For example, we could teach it a generic Reload() bus method call, that we could optionally call from PID 1 instead of ExecReload(), i.e. in this style:

```
[Service]
Type=dbus
BusName=org.foo.service
ReloadByBus=yes
```

I mean, people keep asking us for adding `ReloadSignal=SIGHUP` or so, to allow generic reload implementations via SIGHUP. I never was a fan, since signals are asynchronous, there's no channel for reporting completion back. Method calls otoh have a concept for that, naturally, hence it would make a ton of sense to implement a generic way to reload a service via a Reload() method call, and that could be part of the bus interface we are introducing here.

That all said, it definitely is an option to keep the interface for controlling log settings and the interface for doing Reload() separate. i.e. that fully blown daemons would implement both. org.freedesktop.LogControl1 and org.freedesktop.ServiceLifecycle1 or so, or just one of the two.

I am not entirely sure we need that though? I mean, checking whether a feature is implemented is not better done via checking for interface existence than via checking for method existence...

opinions?

> As for the object, most daemons already have a singleton object that represent "them" and would be a more logical place to to put the interface (from the daemon's point of view, at least)

dbus has this nice concept of allowing many objects on the same service, and i think it is a more natural use of dbus' concepts to have distinct entrypoints for different functionality.

> There is no standard dbus object for "the singleton representing this client", I know... but if such a thing is to be defined, again tying it to the term "service" might not be the best possible name.

I think each application of dbus should define its own entrypoint.

**Comment by poettering** [comment]
> Makes sense. Not declaring a fixed object path and letting this be attached to any path would make it harder to use. As is, we can define `systemd-analyze log-level <service> <level>` easily without any special casing. (Ignoring the fact that "log-level" is used already.)

I'd probably implement this in "busctl" as high-level operation btw. it appears to me this is an (optional) feature of a typical bus service, hence should be exposed in busctl

**Comment by poettering** [comment]
hmm, OK, so gtk introduced the org.freedesktop.Application interface that applications typically implement now. Maybe that's reason to keep the interfaces separate here, i.e. so that people can combine o.fd.A with o.fd.LC1 if they want or o.fd.SL with o.fd.LC1, or maybe all three

Will rework the PR accordingly.

**Comment by poettering** [comment]
Force pushed a new version. No changes, except: renamed the service to org.freedesktop.LogControl1. 

ptal

**Comment by DaanDeMeyer** [comment]
I'd go for "Logging" or "Logger" instead of "LogControl" but that's really just bikeshedding so feel free to keep "LogControl".

I agree with adding a separate object. It seems easier to have a fixed object compared to having to check every object from a service to find the one that implements the interface to modify the log level.

We might want to have a log-level verb on every command line tool as well as busctl. I can imagine somewhat of a frustration having to type out the entire service, object, interface and member when changing the log-level. Having a shorthand available on the respective command line tool takes all that away.

**Comment by poettering** [comment]
@keszybz what's your take on naming this LogControl vs Logging vs something else?

**Comment by poettering** [comment]
(i still have a slight pref for LogControl, but can be convinced otherwise)

**Comment by DaanDeMeyer** [comment]
I realized I didn't provide any arguments. I'd take "Logging" or "Logger" simply because that seems to be what's generally used in other languages/libraries/frameworks/tools. I've never seen LogControl used anywhere else. So basically, I think it would feel more familiar for users if we used the same name used elsewhere to refer to logging stuff.

I suppose we could go even further if we name the interface "Logging"/"Logger" and just name the properties "Level" and "Target" instead of prefixing them with "Log".

Of course, this is all anecdotal and I didn't do an extensive study to see what's used the most.

**Comment by poettering** [comment]
the reason I think LogControl makes sense, since it's not an API for logging. i.e. you cannot issue a log message with it, which is what a logging API probably should do, or what most people would guess a logging API would do. You also cannot read any log messages with it. The API we define here can just control where logging goes, and what is dropped. hence it's really about control, and I think that should be reflected in the name. Does that make sense?

**Comment by dnicolodi** [comment]
A comment from the peanut gallery: I like the naming proposed by @DaanDeMeyer. The exposed methods and properties make clear what this interface is for.

**Comment by keszybz** [comment]
I think the Logging vs LogControl thing is a wash. Let's say that the implementor gets to decide.

**Comment by keszybz** [comment]
bionic-i386 failed with:
```
Fake Lid Switch: /dev/input/event4
Apr 21 16:17:06 autopkgtest systemd-logind[31013]: Watching system buttons on /dev/input/event4 (Fake Lid Switch)
   waiting for suspend, 30s remaining...
   waiting for suspend, 25s remaining...
   waiting for suspend, 20s remaining...
   waiting for suspend, 15s remaining...
   waiting for suspend, 10s remaining...
   waiting for suspend, 5s remaining...
   waiting for suspend, 0s remaining...
closing lid did not cause suspend
autopkgtest [16:17:38]: test logind: -----------------------]
autopkgtest [16:17:38]: test logind:  - - - - - - - - - - results - - - - - - - - - -
logind               FAIL non-zero exit status 1
```
I don't see how this could be related... logind is modified by this PR, but the new interface is not used by anything, so it shouldn't matter.

**Comment by boucman** [comment]
@keszybz what do you mean by "let the implementor decide" ?
If you mean that each daemon should choose a name, that's really not a good idea. The point of interface is to have a well-known name. Everybody should name the interface the same...

(You probably know that, so I'm probably misunderstanding you, but i want to make sure)

**Comment by keszybz** [comment]
I meant the implementor of the patch, i.e. @poettering in this case.

**Comment by boucman** [comment]
ah, ok

**Comment by fbuihuu** [comment]
Nice stuff.

It would great to teach systemctl about the new API and makes `systemctl log-level debug systemd-networkd.service` do the right thing.

**Comment by poettering** [comment]
@fbuihuu i posted #15622 just now which adds something like that to busctl.

**Comment by mbiebl** [comment]
Is the inconsistency in D-Bus interface naming intended? So far we always used lower case, like
`org.freedesktop.{hostname1,import1,locale1,machine1,resolve1,systemd1,timedate1}`.
`org.freedesktop.LogControl1 sticks out here. 
While this is only a cosmetic issue, consistency is always nice.

**Comment by keszybz** [comment]
Yeah, it's done to match this (from the comment near the top):
> so gtk introduced the org.freedesktop.Application interface

**Pull request merged**